src/realm/deppart/sparsity_impl.cc

```diff
@@ -1830,6 +1839,117 @@ namespace Realm {
+            << " entries";
+        for(size_t i = 0; i < this->entries.size(); i++)
+          msg << "\n  [" << i << "]: bounds=" << this->entries[i].bounds
+              << " sparsity=" << this->entries[i].sparsity
+              << " bitmap=" << this->entries[i].bitmap;
+      }
+    }
+
+#ifdef DEBUG_PARTITIONING
+    std::cout << "finalizing " << this << ", " << this->entries.size() << " entries"
+              << std::endl;
+    for(size_t i = 0; i < this->entries.size(); i++)
+      std::cout << "  [" << i << "]: bounds=" << this->entries[i].bounds
+                << " sparsity=" << this->entries[i].sparsity
+                << " bitmap=" << this->entries[i].bitmap << std::endl;
+#endif
+    NodeSet sendto_precise, sendto_approx;
+    Event trigger_precise = Event::NO_EVENT;
+    Event trigger_approx = Event::NO_EVENT;
+    std::vector<PartitioningMicroOp *> precise_waiters_copy, approx_waiters_copy;
+    {
+      AutoLock<> al(mutex);
+
+      assert(!this->entries_valid.load());
+      this->entries_valid.store_release(true);
+
+      precise_requested = false;
+      if(precise_ready_event.exists()) {
+        trigger_precise = precise_ready_event;
+        precise_ready_event = Event::NO_EVENT;
+      }
+
+      precise_waiters_copy.swap(precise_waiters);
+      approx_waiters_copy.swap(approx_waiters);
+
+      remote_precise_waiters.swap(sendto_precise);
+      remote_approx_waiters.swap(sendto_approx);
+    }
+
+    for(std::vector<PartitioningMicroOp *>::const_iterator it =
+            precise_waiters_copy.begin();
+        it != precise_waiters_copy.end(); it++)
+      (*it)->sparsity_map_ready(this, true);
+
+    for(std::vector<PartitioningMicroOp *>::const_iterator it =
+            approx_waiters_copy.begin();
+        it != approx_waiters_copy.end(); it++)
+      (*it)->sparsity_map_ready(this, false);
+
+    if(!sendto_approx.empty()) {
+      for(NodeID i = 0; (i <= Network::max_node_id) && !sendto_approx.empty(); i++)
+        if(sendto_approx.contains(i)) {
+          bool also_precise = sendto_precise.contains(i);
+          if(also_precise)
+            sendto_precise.remove(i);
+          remote_data_reply(i, also_precise, true);
+          sendto_approx.remove(i);
+        }
+    }
+
+    if(!sendto_precise.empty()) {
+      for(NodeID i = 0; (i <= Network::max_node_id) && !sendto_precise.empty(); i++)
+        if(sendto_precise.contains(i)) {
+          remote_data_reply(i, true, false);
+          sendto_precise.remove(i);
+        }
+    }
+
+    if(trigger_approx.exists())
+      GenEventImpl::trigger(trigger_approx, false /*!poisoned*/);
+
+    if(trigger_precise.exists())
+      GenEventImpl::trigger(trigger_precise, false /*!poisoned*/);
+  }
+
+
+  //Allows a GPU deppart client to set the entries directly with a host region instance
+  template<int N, typename T>
+    void SparsityMapImpl<N, T>::set_instance(RegionInstance _entries_instance, size_t size)
```

Comment:
If we're always using `set_instance` from the gpu deppart ops, why do we need to copy/expose a separate `gpu_finalize` function?